File: app/src/main/java/com/treelzebub/umap/ui/fragment/LoginFragment.java
package com.treelzebub.umap.ui.fragment;

import android.app.Fragment;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.treelzebub.umap.R;
import com.treelzebub.umap.api.discogs.DiscogsApi;
import com.treelzebub.umap.api.discogs.DiscogsConstants;

import org.scribe.builder.ServiceBuilder;
import org.scribe.model.Token;
import org.scribe.oauth.OAuthService;

import butterknife.ButterKnife;
import butterknife.InjectView;
import butterknife.OnClick;

/**
 * Created by Tre Murillo on 3/23/15
 * <p/>
 * A fragment which will eventually provide a one-time login to Discogs (and optionally, Gemm)
 */
public class LoginFragment extends Fragment {

    public static final String TAG = "LoginActivity";

    private String authUrl;

    @InjectView(R.id.webview) WebView        mWebView;
    @InjectView(R.id.auth_code_et) EditText  mAuthCodeET;
    @InjectView(R.id.submit_button) Button   mSubmitButton;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                OAuthService service = new ServiceBuilder()
                        .apiKey(DiscogsConstants.CONSUMER_KEY)
                        .apiSecret(DiscogsConstants.CONSUMER_SECRET)
                        .callback(DiscogsConstants.CALLBACK_URL)
                        .provider(DiscogsApi.class)
                        .build();
                Token rt = service.getRequestToken();
                authUrl = service.getAuthorizationUrl(rt) + DiscogsConstants.AUTH_URL_APPEND + "<" + rt.getToken() + ">";
                return null;
            }

            @Override
            protected void onPostExecute(Void foo) {
                mWebView.loadUrl(authUrl);
            }
        }.execute();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View v = inflater.inflate(R.layout.dialog_oauth, container, false);
        ButterKnife.inject(this, v);
        return v;
    }

    @OnClick(R.id.submit_button)
    protected void submit(View v) {
        CharSequence authCode = mAuthCodeET.getText();
        if (authCode.length() > 0) {
            Intent intent = new Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse(DiscogsConstants.BASE_URL + "/login" + "?client_id=" + DiscogsConstants.CONSUMER_KEY + "&redirect_uri=" + DiscogsConstants.CALLBACK_URL));
            startActivity(intent);
            // GOAL!
        } else {
            Toast.makeText(getActivity(), "Please enter authorization code", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/kotlin/com/treelzebub/umap/ui/LoginFragment.kt
package com.treelzebub.umap.ui

import android.content.Intent
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.webkit.WebViewClient
import butterknife.bindView
import com.treelzebub.umap
import com.treelzebub.umap.DISCOGS_AUTH_URL_APPEND
import com.treelzebub.umap.DISCOGS_CONSUMER_KEY
import com.treelzebub.umap.DISCOGS_CONSUMER_SECRET
import com.treelzebub.umap.R
import com.treelzebub.umap.auth.DiscogsApi
import com.treelzebub.umap.auth.TokenHolder
import com.treelzebub.umap.util.BusProvider
import org.scribe.builder.ServiceBuilder

/**
 * Created by Tre Murillo on 5/28/15
 *
 * A fragment that provides a one-time login to Discogs.com
 */
public class LoginFragment : Fragment() {

    companion object {
        private val TAG = javaClass<LoginFragment>().getSimpleName()
    }

    private var authUrl: String? = null

    val webView: WebView by bindView(R.id.webview)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        BusProvider.instance.register(this)
        loadAuthUrl()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.fragment_login, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        webView.getSettings().setBuiltInZoomControls(true)
        webView.getSettings().setJavaScriptEnabled(true)
        webView.setWebViewClient(Callback())
    }

    private fun loadAuthUrl() {
        object : AsyncTask<Void, Void, Void>() {
            override fun doInBackground(vararg params: Void): Void? {
                val oAuthService = ServiceBuilder()
                        .apiKey(DISCOGS_CONSUMER_KEY)
                        .apiSecret(DISCOGS_CONSUMER_SECRET)
                        .callback(umap.CALLBACK_URL)
                        .provider(javaClass<DiscogsApi>())
                        .build()
                val rt = oAuthService.getRequestToken()
                TokenHolder.requestToken = rt
                authUrl = oAuthService.getAuthorizationUrl(rt) + DISCOGS_AUTH_URL_APPEND + rt.getToken()
                return null
            }

            override fun onPostExecute(result: Void?) {
                webView.loadUrl(authUrl)
            }
        }.execute()
    }

    private inner class Callback : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String?): Boolean {
            if (url != null && url.startsWith(umap.CALLBACK_URL)) {
                val i = Intent(getActivity(), javaClass<DashboardActivity>()).setData(Uri.parse(url))
                startActivity(i)
                return true
            }
            return false
        }
    }
}
